### ISSUE TITLE
drop --reckless kill   ignores subdatasets under folders pointed to by `drop` command

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#7013]
It might be something obvious but I am missing it: do we have a way to invoke `uninstall` or `drop` to uninstall/drop all subdatasets but leave actual dataset behind?  I have tried

```
(dandisets) dandi@drogon:/mnt/backup/dandi/dandisets/000108$ datalad drop --what datasets --reckless kill -r -J10 
[ERROR  ] refuse to perform actions that would remove the current working directory (RuntimeError) 
(dandisets) dandi@drogon:/mnt/backup/dandi/dandisets/000108$ datalad drop --what datasets --reckless kill -r -J10 .
[ERROR  ] refuse to perform actions that would remove the current working directory (RuntimeError) 
```

so "current one" gets in the way, limiting to subfolders doesn't work

```
(dandisets) dandi@drogon:/mnt/backup/dandi/dandisets/000108$ datalad drop --what datasets --reckless kill -r -J10 sub*
```
in that it does nothing really. So the only way I see is to explicitly glob all subdatasets I know about:

```
(dandisets) dandi@drogon:/mnt/backup/dandi/dandisets/000108$ datalad drop --what datasets --reckless kill -r -J10 ./sub-*/ses-*/*/*.ome.zarr
uninstall(ok): sub-MITU01h3/ses-20220313h17m18s08/micr/sub-MITU01h3_ses-20220313h17m18s08_sample-25_stain-SST_run-1_chunk-4_SPIM.ome.zarr (dataset)
uninstall(ok): sub-MITU01h3/ses-20220313h17m18s08/micr/sub-MITU01h3_ses-20220313h17m18s08_sample-25_stain-SST_run-1_chunk-3_SPIM.ome.zarr (dataset)
...
```

and then it "goes" but BTW **very slowly** with 100% CPU by datalad (0.17.2, something to figure out I guess, might relate to #6974 but unlikely since it is how slow it goes after each subdataset, not some initial analysis),

**Comment by mih** [comment]
Not sure what is going on here:

```sh
(datalad-dev) mih@meiner /tmp % cd top
(datalad-dev) mih@meiner /tmp/top (git)-[master] % datalad create -d . sub1
add(ok): sub1 (dataset)                                                                                          
add(ok): .gitmodules (file)                                                                                      
save(ok): . (dataset)                                                                                            
create(ok): sub1 (dataset)                                                                                       
action summary:
  add (ok: 2)
  create (ok: 1)
  save (ok: 1)
(datalad-dev) mih@meiner /tmp/top (git)-[master] % datalad create -d . subdir/sub2
add(ok): subdir/sub2 (dataset)                                                                                   
add(ok): .gitmodules (file)                                                                                      
save(ok): . (dataset)                                                                                            
create(ok): subdir/sub2 (dataset)                                                                                
action summary:
  add (ok: 2)
  create (ok: 1)
  save (ok: 1)
(datalad-dev) mih@meiner /tmp/top (git)-[master] % datalad drop --reckless availability -r --what datasets *
uninstall(ok): sub1 (dataset)
uninstall(ok): subdir/sub2 (dataset)
action summary:
  uninstall (ok: 2)

```

seems to work as desired.

**Comment by yarikoptic** [comment]
The culprit is `--reckless kill` which I used, and you @mih use `--reckless availability`. With `--reckless kill` it seems to not just disable all safety checks but manages to not perform operation at all silently. Here is a reproducer/example:

```
#!/bin/bash

export PS4='> '
set -x
set -eu
cd "$(mktemp -d ${TMPDIR:-/tmp}/dl-XXXXXXX)"


datalad create .
mkdir d1

# in subfolder
datalad install -d . -s  ///dandi d1/sub
# on the top
datalad install -d . -s  ///dandi sub

datalad subdatasets --state present

datalad drop --what all --reckless kill -r *
datalad subdatasets --state present

datalad drop --what all --reckless availability -r *
datalad subdatasets --state present
```

in which I create two subdatasets -- one on top level and on in subfolder `d1/` .  `--reckless kill` drops only the top level but leaves the one in subfolder behind. the `--reckless availability` completes operation correctly:

```
...
> datalad subdatasets --state present
subdataset(ok): d1/sub (dataset)
subdataset(ok): sub (dataset)
> datalad drop --what all --reckless kill -r d1 sub
uninstall(ok): sub (dataset)
action summary:
  drop (notneeded: 1)
  uninstall (ok: 1)
> datalad subdatasets --state present
subdataset(ok): d1/sub (dataset)
> datalad drop --what all --reckless availability -r d1 sub
uninstall(ok): d1/sub (dataset)
action summary:
  drop (notneeded: 3)
  uninstall (ok: 1)
> datalad subdatasets --state present
bash create-drop-kill.sh  6.71s user 2.68s system 98% cpu 9.532 total
```

I am adjusting the title.
```

**Comment by bpoldrack** [comment]
Reproduced. Given the difference with `kill` and `availability` I consider this a bug.

**Comment by adswa** [comment]
I've looked at this a bit. Here is how the problem unfolds:

``get_paths_by_ds()`` of the paths ``['d1', 'sub'] returns 
```
{
    PosixPath('/tmp/dl-d3vnZfY'): [PosixPath('/tmp/dl-d3vnZfY/d1')], 
    PosixPath('/tmp/dl-d3vnZfY/sub'): [PosixPath('/tmp/dl-d3vnZfY/sub')]
}
```
It then performs a ``drop`` on each dataset - the top-level one, and the ``sub`` subdataset. This operation succeeds for the ``sub`` dataset, but doesn't as the following code modifies ``what`` to ``filecontent``: The dataset object is the toplevel one, the path is a directory underneath. This makes it limit the to-be-dropped-stuff to only file content:

```
    if paths is not None and paths != [ds.pathobj] and what == 'all':
        # so we have paths constraints that prevent dropping the full dataset
        lgr.debug('Only dropping file content for given paths in %s, '
                  'allthough instruction was to drop %s', ds, what)
        what = 'filecontent'
``` 

The code also wouldn't remove all subdatasets  without the path constraint - it would error out since it would also need to do a removal of the worktree. 

The following patch might do:

```diff
diff --git a/datalad/distributed/drop.py b/datalad/distributed/drop.py
index 5141148e2..9efe50f10 100644
--- a/datalad/distributed/drop.py
+++ b/datalad/distributed/drop.py
@@ -346,10 +346,30 @@ def _drop_dataset(ds, paths, what, reckless, recursive, recursion_limit, jobs):
         return
 
     if paths is not None and paths != [ds.pathobj] and what == 'all':
-        # so we have paths constraints that prevent dropping the full dataset
-        lgr.debug('Only dropping file content for given paths in %s, '
-                  'allthough instruction was to drop %s', ds, what)
-        what = 'filecontent'
+        ds_updated = False
+        if recursive and reckless == 'kill':
+            # check if any paths contains a subdataset
+            for sub in ds.subdatasets(
+                    # must be resolved!
+                    path=paths,
+                    state='present',
+                    recursive=recursive,
+                    recursion_limit=recursion_limit,
+                    # start reporting with the leaves
+                    bottomup=True,
+                    result_xfm='datasets',
+                    on_failure='ignore',
+                    return_type='generator',
+                    result_renderer='disabled'):
+                if sub is not None:
+                    # there is a subdataset underneath the given path
+                    ds_updated = True
+                    ds = sub
+        if not ds_updated:
+            # so we have paths constraints that prevent dropping the full dataset
+            lgr.debug('Only dropping file content for given paths in %s, '
+                      'allthough instruction was to drop %s', ds, what)
+            what = 'filecontent'
 
     repo = ds.repo
     is_annex = isinstance(repo, AnnexRepo)
```

**Comment by adswa** [comment]
Fixed in #7308
